File: 76.minimum-window-substring.cpp
/*
 * @lc app=leetcode id=76 lang=cpp
 *
 * [76] Minimum Window Substring
 */

// @lc code=start
#include <string>
#include <unordered_map>

class Solution {
public:
    std::string minWindow(std::string s, std::string t) {
        std::unordered_map<char, int> need;
        int missing = t.length(), i = 0, I = -1, J = -1;

        for (char c : t) {
            if (need.find(c) != need.end()) {
                need[c]++;
            } else {
                need[c] = 1;
            }
        }

        for (int j = 0; j < s.length(); j++) {
            char c = s[j];
            if (need.find(c) == need.end()) need[c] = 0;
            if (need[c] > 0) {
                missing--;
            }
            need[c]--;

            if (missing == 0) {
                while (i < j && (need.find(s[i]) != need.end() && need[s[i]] < 0)) {
                    need[s[i]]++;
                    i++;
                }

                if (J == -1 || j - i <= J - I) {
                    I = i;
                    J = j;
                }
            }
        }

        if (missing == 0) {
            return s.substr(I, J - I + 1);
        } else {
            return "";
        }
    }
};

// @lc code=end

File: 76.minimum-window-substring.py
#
# @lc app=leetcode id=76 lang=python3
#
# [76] Minimum Window Substring
#

# @lc code=start
import collections


class Solution:
    '''
    # First count occurence of characters in t
    # Then for each index j, ask if s[i:j + 1] satisfy the substring requirement
    # need[c] can become negative, which means we have excess of the current character or it does not appear in t
    # missing = number of not yet appeared characters that is present in t
    '''
    def minWindow(self, s: str, t: str) -> str:
        need, missing, i, I, J = collections.Counter(t), len(t), 0, -1, -1
        for j, c in enumerate(s):
            if need[c] > 0:
                missing -= 1
            need[c] -= 1
            if missing == 0:
                while i < j and need[s[i]] < 0:
                    need[s[i]] += 1
                    i += 1
                if J == -1 or j - i <= J - I:
                    I, J = i, j
        if missing == 0:
            return s[I:J+1]
        else:
            return ""
# @lc code=end

